walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,202 @@
+   * @param request MeetingRequest object which contains list of attendees
+   * @param event The Event object's attendees are compare with the request object
+   * @return boolean if the provide event object contains no mandatory attendees
+   */
+  private boolean eventNoMandatoryAttendees(MeetingRequest request, Event event){
+    for (String attendee : event.getAttendees()){
+      if (request.getAttendees().contains(attendee)){
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * This method checks if an event contains only optional Attendees using a MeetingRequest
+   * @param request MeetingRequest object which contains list of optional attendees
+   * @param event The Event object's attendees are compare with the request object
+   * @return boolean if the provide event object contains only optional attendees
+   */
+  private boolean optionalAttendeesEvent(MeetingRequest request, Event event){
+    for (String attendee : event.getAttendees()){
+      if (request.getAttendees().contains(attendee)){
+        return false;
+      }
+      if (!(request.getOptionalAttendees().contains(attendee))){
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * This method returns the correct time ranges if optional attendees are taken into account
+   * @param optionalEvents List of events with only optional attendees
+   * @param timeRanges A list of available TimeRanges without taken into account optional events 
+   * @return A list of TimeRanges depending on the optionalEvents list and final size of timeRanges
+   */
+  private Collection<TimeRange> resolveOptionalEventConflicts(ArrayList<Event> optionalEvents, Collection<TimeRange> timeRanges){
+    Collection<TimeRange> backUpTimeRanges = new ArrayList<TimeRange>(timeRanges);
+    ArrayList<TimeRange> removedTimeRanges = new ArrayList<TimeRange>();
+
+    for (Event opEvent : optionalEvents){
+      for (TimeRange timeRange : timeRanges){
+        if (opEvent.getWhen().overlaps(timeRange)){
+          removedTimeRanges.add(timeRange);
+        }
+      }
+    }
+
+    for (TimeRange removedTimeRange : removedTimeRanges){
+      timeRanges.remove(removedTimeRange);
+    }
+
+    if (timeRanges.isEmpty()){
+      return backUpTimeRanges;
+    }
+    return timeRanges;
+  }
+
+  /**
+   * This method checks if the previous and current time ranges are nested events
+   * @param prevTimeRange Previous TimeRange object
+   * @param currTimeRange Current TimeRange object
+   * @return true if the TimeRange objects meet the criteria of nested events
+   */
+  private boolean nestedEventsCheck(TimeRange prevTimeRange, TimeRange currTimeRange){
+    if(prevTimeRange.start() < currTimeRange.start() && currTimeRange.end() < prevTimeRange.end()){
+      return true;
+    }
+    return false;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```

Comment:
This is a very large method with several loops and a bunch of conditionals. See if you can split it up to use a few private methods to help with readability.

Adding some documentation to this method would be helpful too.

Reply:
Ok, I added the ```processOptionalEvents``` function to the algorithm to hopefully increase readability. Sorry for the long ```query``` function.